heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java

```diff
@@ -296,6 +303,13 @@ public QueryTrace buildTrace() {
 
         @Override
         public AsyncFuture<FullQuery> query(final FullQuery.Request request) {
+            if (!concurrentQueries.tryAcquire()) {
+                // There's currently too many concurrent queries. Fail now so that the QueryManager
+                // gets an opportunity to try another node in the same shard instead.
+                throw new GoAwayException(
+                    "Node has reached maximum number of concurrent MetricManager requests (" +
+                        concurrentQueriesBackoff + ")");
+            }
```

Comment:
Any synchronous exception in this code block will cause the semaphore to leak leases, and eventually lock up.

I'd suggest using a helper function that catches these, and releases the semaphore for but circumstances.
* An immediate exception is thrown.
* The future completes.

Reply:
Can't believe I missed that 😊  Thanks

Reply:
Fixed this now, in addition to spotting an issue that existed previously also: A non thread safe list being used for keeping track of which nodes that had been tried. Introduced by me when working on retry code some time ago. Fixed now :)